### PR TITLE
[WIP] Adds Dockerfile

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
This PR aims to add a Dockerfile for the `poutyne`. 

@freud14 while this image gets successfully built, I'd like to further tone down the image size. At the moment it's about `6.35 GB` which makes no sense as the base image itself is `2.79 GB`. After using [**`dive`**](https://gochronicles.com/dive/) to inspect the image it turns out the `COPY` statement from the base image takes up around `~6.2GB`.

I've never encountered this problem before, would be grateful if somebody else pitches in.

To build the 🐳 image use `DOCKER_BUILDKIT=1 docker build -t poutyne .`

### CODE REVIEW

**Comment by freud14** [comment]
Hi @SauravMaheshkar, according to the docker hub website, the image [`pytorch/pytorch:1.10.0-cuda11.3-cudnn8-runtime`](https://registry.hub.docker.com/layers/pytorch/pytorch/1.10.0-cuda11.3-cudnn8-devel/images/sha256-913e6689c5958b187e65561e528ec6c3ce8a02deedcdd38cb50c9cab301907bb?context=explore#!) is supposed to be more than 6 GB. So your numbers make sense. This is also what I get:
docker-poutyne
What do you think?

**Comment by SauravMaheshkar** [comment]
Huh, Yes. Don't know why I was confused. Converting this to be ready for Review. 

Any other thoughts on the `Dockerfile` ?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    && find /opt/conda/lib/ -name '*.pxd' -delete \
+    && find /opt/conda/lib/ -follow -type f -name '*.md' -delete \
+    && find /opt/conda/lib/ -follow -type f -name '*.png' -delete \
+    && find /opt/conda/lib/ -follow -type f -name '*.jpg' -delete \
+    && find /opt/conda/lib/ -follow -type f -name '*.jpeg' -delete \
+    && find /opt/conda/lib/ -name '*.pyd' -delete \
+    && find /opt/conda/lib/ -name '__pycache__' | xargs rm -r
+
+ENV PATH /opt/conda/bin:$PATH
+
```
Comment (freud14) [inline Dockerfile:33]:
```suggestion
```
There are two empty lines at the end of the file.

**Comment by freud14** [comment]
Hi, two things:
* You used the dev branch in the Dockerfile. Is it on purpose? I am not a frequent user of Docker so I am not sure of the common use cases.
* How do you use this kind of Dockerfiles when developing? Do you think it could be a good idea to add some instructions in the README and on the documentation website?

**Comment by freud14** [comment]
@SauravMaheshkar Any news about that?

**Comment by freud14** [comment]
> * I used the dev branch in the `Dockerfile`, to have the latest version of the library installed in the image.

If it is to be used for deployment, wouldn't it be preferable to use the stable version? Is it possible to put flags (like a dev flag) ?

> * It could be a deep-dive tutorial into how one can use the image to do deployment.

I think this one is pretty straightforward since someone wanting to deploy an application would "inherit" (sorry I don't know all the vocab) the Dockerfile and just add its own content.

>       * It could be an alternative way to work with the repository. For instance instead of pip or conda one can just use the image to prototype and develop.

I think this one might be preferable.

> To add on, it might also be useful to add some CI/CD regarding the image

Right now, we're using GitHub Actions for CI. What are you proposing exactly?

**Comment by SauravMaheshkar** [comment]
- Yes, makes sense, I'll edit the file to use the stable version. We can use flags but that would involve having multiple images be viz. `poutyne:latest` and `poutyne:stable`. But I'd suggest not to have multiple `Dockerfile`'s. We should just have the "stable" `Dockerfile` and manually release the other stable versions of the library.
- So do you want me to add instructions about this `Dockerfile` in the installation section ?
- I'm proposing adding Github Action Workflows for the docker image. Some workflows that I personally use are :-

<details>
<summary> Build Docker 🐳  Image </summary>

```yaml

name: Docker Image CI

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  release:
    types: [created]
  schedule:
    # Run the tests at 00:00 each day
    - cron: "0 0 * * *"

jobs:

  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Build the Docker image
      run: |
        docker build . --file Dockerfile --tag poutyne-dev:$(date +%s)

```


 </details>

<details>
<summary> Publish Docker 🐳  Image to Github Container Registry</summary>

```yaml

name: Create and publish a Docker image as a Github Package

on:
  release:
    types: [created]

env:
  REGISTRY: ghcr.io
  IMAGE_NAME: ${{ github.repository }}-dev

jobs:
  build-and-push-image:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write

    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Log in to the Container registry
        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
        with:
          registry: ${{ env.REGISTRY }}
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}

      - name: Extract metadata (tags, labels) for Docker
        id: meta
        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
        with:
          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}

      - name: Build and push Docker image
        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
        with:
          context: .
          push: true
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}

```


 </details>

**Comment by freud14** [comment]
> * Yes, makes sense, I'll edit the file to use the stable version. We can use flags but that would involve having multiple images be viz. `poutyne:latest` and `poutyne:stable`. But I'd suggest not to have multiple `Dockerfile`'s. We should just have the "stable" `Dockerfile` and manually release the other stable versions of the library.

Alright.

> * So do you want me to add instructions about this `Dockerfile` in the installation section ?

Yes, that would be nice. Something short.

> Build Docker whale Image

That seems nice. Not sure I see the point of the cron job. Is it in case the base PyTorch image changes?

> Publish Docker whale Image to Github Container Registry

Seems nice too. Why is the image name suffixed by "-dev" if the image is created on release?

**Comment by SauravMaheshkar** [comment]
> That seems nice. Not sure I see the point of the cron job. Is it in case the base PyTorch image changes?

Yes, that's the edge case that the cron job targets. Although not likely, we can skip if you think need be.

> Seems nice too. Why is the image name suffixed by "-dev" if the image is created on release?

Nothing in particular, I just wrote `dev` in this chat to signify it's under development.

**Comment by freud14** [comment]
Hi, it seems that the merge is weird in Github. Do you think you could fix that?

**Comment by freud14** [comment]
I mean in the "Files changed" tab where we see the modifications of all commits.

**Comment by freud14** [comment]
Worst case, could you make another PR but starting it from the dev branch instead of master? Otherwise, it becomes hard to review.

**Comment by freud14** [comment]
Hi @SauravMaheshkar, I created another PR (#153) from your PR. If you have time, let me know what you think.

**Pull request closed without merging**